va/va.go

```diff
@@ -29,96 +39,177 @@ func userAgent() string {
+
+		// Update the validated date for the challenge regardless of good/bad outcome
+		now := va.clk.Now().UTC()
+		chal.ValidatedDate = now
+		chal.Validated = chal.ValidatedDate.Format(time.RFC3339)
+
+		// Read a validation result from the task results channel for each of the
+		// concurrent validations into a slice for processing.
+		for i := 0; i < concurrentValidations; i++ {
+			va.log.Printf("Reading a result from task.Results")
+			result := <-task.Results
+
+			// If one of the results was an error, the challenge fails
+			if result.Error != nil {
+				// Update the challenge Error
+				chal.Error = result.Error
+				// Set the challenge and authorization to invalid
+				chal.Status = acme.StatusInvalid
+				authz.Status = acme.StatusInvalid
+				va.log.Printf("authz %s set INVALID by completed challenge %s", authz.ID, chal.ID)
+				// Continue immediately, the challenge is finished
+				continue TaskLoop
+			}
+		}
+
+		// If none of the results were an error then the challenge succeeded.
 		// Update the expiry for the valid authorization
 		authz.ExpiresDate = now.Add(validAuthzExpire)
 		authz.Expires = authz.ExpiresDate.Format(time.RFC3339)
 		// Set the authorization & challenge to valid
 		chal.Status = acme.StatusValid
 		authz.Status = acme.StatusValid
 		va.log.Printf("authz %s set VALID by completed challenge %s", authz.ID, chal.ID)
+
+		// Check whether validating this authorization completed the overall order
+		// TODO(@cpu): this will race if another thread updates an authorization for this order
+		order := authz.Order
+		for _, authz := range order.AuthorizationObjects {
+			// If any of the authorizations are invalid the order isn't ready to issue
+			if authz.Status != acme.StatusValid {
+				continue TaskLoop
+			}
+		}
+		// Ask the CA to complete the order in a separate goroutine
+		go va.ca.CompleteOrder(order)
 	}
+}
+
+func (va VAImpl) performValidation(task *vaTask, wg *sync.WaitGroup) {
+	// Sleep for a random amount of time between 1-15s
+	len := time.Duration(rand.Intn(15))
+	va.log.Printf("Sleeping for %s seconds before validating", time.Second*len)
+	va.clk.Sleep(time.Second * len)
 
-	return nil
+	// TODO(@cpu): Implement validation for DNS-01, TLS-SNI-02, etc
+	switch task.Challenge.Type {
+	case acme.ChallengeHTTP01:
+		va.validateHTTP01(task)
+	default:
+		va.log.Printf("Error: performValidation(): Invalid challenge type: %q", task.Challenge.Type)
+	}
+	va.log.Printf("Finished a performValidation()")
+	wg.Done()
 }
 
-func (va VAImpl) validateHTTP01(
-	identifier string,
-	chal *core.Challenge,
-	reg *core.Registration,
-) *acme.ProblemDetails {
-	body, err := va.fetchHTTP(identifier, chal)
+func (va VAImpl) validateHTTP01(task *vaTask) {
+	va.log.Printf("Validating HTTP01")
+	body, url, err := va.fetchHTTP(task.Identifier, task.Challenge.Token)
+
+	result := core.ValidationRecord{
+		URL:         url,
+		ValidatedAt: va.clk.Now(),
+	}
 	if err != nil {
-		return err
+		result.Error = err
```

Comment:
I think switching this function from one that returns an error, to one that mutates its input to set an error, is a step backwards in clarity. We had similar code in Boulder, but moved towards the "return an error" approach.

Stepping back a bit, this is related to the choice to have a `vaTask` contain both input and output. What if we defined `vaTask` to be only the input to a task, and the output from a task to be a ValidationRecord? Or in other words, treat vaTask as immutable. This might be an argument to use a `chan *ValidationRecord` in place of a `WaitGroup`. The `wg` argument to `performValidation` could be replaced with a `chan`, and the switch statement could do `ch <- va.validateHTTP01(task)`.